src/ui/pages/SwapForm/SwapForm.tsx

```diff
@@ -166,387 +172,411 @@ function FormHint({
   return render(hint);
 }
 
-export function SwapFormComponent() {
-  useBackgroundKind({ kind: 'white' });
-  const { singleAddress: address, ready } = useAddressParams();
-  const { currency } = useCurrency();
+function getDefaultState({
+  address,
+  positions,
+}: {
+  address: string;
+  positions: AddressPosition[] | null;
+}): SwapFormState {
+  return { inputChain: getDefaultChain(address, positions ?? []) };
+}
 
-  const { data: wallet } = useQuery({
-    queryKey: ['wallet/uiGetCurrentWallet'],
-    queryFn: () => walletPort.request('uiGetCurrentWallet'),
-    useErrorBoundary: true,
+async function queryPopularTokens(chain: Chain) {
+  return queryClient.fetchQuery({
+    queryKey: ['getPopularTokens', chain],
+    queryFn: () => getPopularTokens(chain),
+    staleTime: Infinity,
   });
+}
 
-  const { data: positionsResponse } = useHttpAddressPositions(
+/**
+ * This value must be used until `prepareDefaultState` resolves
+ * so that `formState` object has stable order of keys. This is important
+ * for useQuotes helper which internally serializes formState into a URL string
+ * and makes refetches when this url changes
+ */
+const EMPTY_DEFAULT_STATE = {
+  inputChain: undefined,
+  inputFungibleId: undefined,
+  outputFungibleId: undefined,
+};
+async function prepareDefaultState({
+  address,
+  userStateInputChain,
+  currency,
+}: {
+  address: string;
+  userStateInputChain: string | null;
+  currency: string;
+}): Promise<SwapFormState> {
+  const source = await getHttpClientSource();
+  const { data: allPositions } = await queryHttpAddressPositions(
```

Comment:
As I can see, we have positions in the parent component. What is better here - pass them or fetch them separately?

Reply:
The point of using QueryClient is that we don't need to worry about this :)